crypto/x509/x509_vfy.c

```diff
@@ -1743,37 +1725,24 @@ static int internal_verify(X509_STORE_CTX *ctx)
 {
     int n = sk_X509_num(ctx->chain) - 1;
     X509 *xi = sk_X509_value(ctx->chain, n);
-    X509 *xs;
+    X509 *xs = xi;
 
     /*
      * With DANE-verified bare public key TA signatures, it remains only to
      * check the timestamps of the top certificate.  We report the issuer as
      * NULL, since all we have is a bare key.
      */
     if (ctx->bare_ta_signed) {
-        xs = xi;
         xi = NULL;
         goto check_cert_time;
     }
 
-    if (ctx->check_issued(ctx, xi, xi))
-        xs = xi; /* the typical case: last cert in the chain is self-issued */
-    else {
-        if (ctx->param->flags & X509_V_FLAG_PARTIAL_CHAIN) {
-            xs = xi;
-            goto check_cert_time;
-        }
-        if (n <= 0) {
-            CHECK_CB(1, ctx, xi, 0, X509_V_ERR_UNABLE_TO_VERIFY_LEAF_SIGNATURE);
```

Comment:
If the chain building yields a chain where the last cert in the chain is not self-issued (regardless whether it's an EE or CA cert), this can only happen if that cert is trusted and X509_V_FLAG_PARTIAL_CHAIN is active, right?
Then there is no point in trying to verify its signature, as I wrote in the new comment.

Reply:
Well, chain building can also fail to construct a trusted chain.  This callback was raising an error that the application verify callback could optionally override. That error does not seem to be raised anymore, so we need to be sure that's OK.

Reply:
In case the chain building cannot find a path that ends in a trust anchor it must not declare success.
So the chain verification (and thus the trigger for that callback) should not be reachable at all in this case!

Reply:
BTW, to be on the safe side I advise to make `internal_verify()` (double-)check that the end of the chain presented to it is trusted. This should be done not only in the special case of a not self-issued cert but generally. This improvement is proposed in #13770.

Reply:
That might break DANE with bare public key, though I probably have tests for that.  Caution advised.  And the check is not entirely free, I am not sure it is worth it.

Reply:
I have essentially zero knowledge of DANE, so would be great if you can check the DANE implications - maybe the approach needs to adapted to this end.

Reply:
Yes, tests `#44` and `#45` in `test/danetest.in` check bare public key trust-anchor support.

Reply:
And the tests still pass :)